src/coreComponents/physicsSolvers/fluidFlow/ImmiscibleMultiphaseFlow.cpp

```diff
@@ -0,0 +1,1277 @@
+      if( m_hasCapPressure )
+      {
+        arrayView2d< real64 const > const porosity = porousMaterial.getPorosity();
+
+        string const & permName = subRegion.getReference< string >( viewKeyStruct::permeabilityNamesString() );
+        PermeabilityBase const & permeabilityMaterial =
+          getConstitutiveModel< PermeabilityBase >( subRegion, permName );
+        arrayView3d< real64 const > const permeability = permeabilityMaterial.permeability();
+
+        string const & capPressName = subRegion.getReference< string >( viewKeyStruct::capPressureNamesString() );
+        CapillaryPressureBase const & capPressureMaterial =
+          getConstitutiveModel< CapillaryPressureBase >( subRegion, capPressName );
+        capPressureMaterial.saveConvergedRockState( porosity, permeability );
+      }
+    } );
+  } );
+}
+
+void ImmiscibleMultiphaseFlow::saveConvergedState( ElementSubRegionBase & subRegion ) const
+{
+  FlowSolverBase::saveConvergedState( subRegion );
+
+  arrayView2d< real64 const, immiscibleFlow::USD_PHASE > const phaseVolFrac =
+    subRegion.getField< fields::immiscibleMultiphaseFlow::phaseVolumeFraction >();
+  arrayView2d< real64, immiscibleFlow::USD_PHASE > const phaseVolFrac_n =
+    subRegion.getField< fields::immiscibleMultiphaseFlow::phaseVolumeFraction_n >();
+  phaseVolFrac_n.setValues< parallelDevicePolicy<> >( phaseVolFrac );
+
+  arrayView2d< real64 const, immiscibleFlow::USD_PHASE > const phaseMass =
+    subRegion.getField< fields::immiscibleMultiphaseFlow::phaseMass >();
+  arrayView2d< real64, immiscibleFlow::USD_PHASE > const phaseMass_n =
+    subRegion.getField< fields::immiscibleMultiphaseFlow::phaseMass_n >();
+  phaseMass_n.setValues< parallelDevicePolicy<> >( phaseMass );
+
+}
+
+void ImmiscibleMultiphaseFlow::updateState( DomainPartition & domain )
+{
+  GEOS_MARK_FUNCTION;
+
+  forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&]( string const &,
+                                                               MeshLevel & mesh,
+                                                               arrayView1d< string const > const & regionNames )
+  {
+    mesh.getElemManager().forElementSubRegions< CellElementSubRegion,
+                                                SurfaceElementSubRegion >( regionNames, [&]( localIndex const,
+                                                                                             auto & subRegion )
+    {
+      // update porosity, permeability, and solid internal energy
+      updatePorosityAndPermeability( subRegion );
+      // update all fluid properties
+      updatePhaseVolumeFraction( subRegion );
+      updateFluidState( subRegion );
+    } );
+  } );
+}
+
+real64 ImmiscibleMultiphaseFlow::setNextDt( const geos::real64 & currentDt, geos::DomainPartition & domain )
+{
+  return PhysicsSolverBase::setNextDt( currentDt, domain );
```

Comment:
TODO pressure and saturation change time step selector

Reply:
Implemented new function:
`real64 ImmiscibleMultiphaseFlow::setNextDtBasedOnStateChange(real64 const & currentDt, DomainPartition & domain) `